numpy/core/src/npysort/npysort_common.h

```diff
@@ -329,7 +329,7 @@ UNICODE_LT(npy_ucs4 *s1, npy_ucs4 *s2, size_t len)
 NPY_INLINE static int
 DATETIME_LT(npy_datetime a, npy_datetime b)
 {
-    return a < b;
+    return a < b ? a != NPY_DATETIME_NAT : b == NPY_DATETIME_NAT;
```

Comment:
I don't think this is correct. If both `a` and `b` are NaT, then the return value should be False, but here is will be True.

Reply:
Do we have an idiomatic way to probe sort stability of identical array elements at the Python level? I want a test to fail if this is wrong.

For example, the following isn't good enough to detect an issue, but may gave you an idea of what I'm thinking about:

```python
 139     def test_lt_nat_stability(self):
 140         # NaT < NaT should be False internally for sort purposes
 141         arr = np.array(['NaT', 'NaT'], dtype='M8')
 142         ids_expected = id(arr[0]), id(arr[1])
 143         arr.sort()
 144         ids_actual = id(arr[0]), id(arr[1])
 145         assert_equal(ids_actual[0], ids_expected[0])
```

Does anything (pointer, address number exposed somewhere else) change if they swap?

Reply:
You cannot tell just sorting the elements, so we rely on argsort. An array of `Nat` should do the trick if you use a stable sort, but you do need different lengths as mergesort fall back on insertion sort for smaller arrays. There are some tests in `test_multiarray.py`, search for `# check that stable argsorts are stable`, but is looks incomplete to me as it should also test arrays of `NaN` and `NaT`. By current standards, the tests aren't all that complete :(

Reply:
I was going to recommend sorting a structured array with one element the time with `NaT`, and another effectively an index, but `argsort` is more direct (as long as we're sure it follows the same sorting path).

Reply:
> as long as we're sure it follows the same sorting path

It should, but of course there could be bugs :) What we are really testing is the `*_LT` macro. I think there are also some tests for NaNs, but I didn't notice any for NaNs and stability combined.